### PR TITLE
Fix missing flush in `testReadBlobWithPrematureConnectionClose`

### LINKED ISSUE #113595: [CI] URLBlobContainerRetriesTests testReadBlobWithPrematureConnectionClose failing
**Build Scans:**
- [elasticsearch-periodic #4201 / openjdk23_java-matrix](https://gradle-enterprise.elastic.co/s/ox2rdcz4v4kc6)
- [elasticsearch-periodic #4193 / openjdk23_java-matrix](https://gradle-enterprise.elastic.co/s/a4tk64oyfy5ju)
- [elasticsearch-periodic #4189 / openjdk23_java-matrix](https://gradle-enterprise.elastic.co/s/kylg5jatnnpha)
- [elasticsearch-periodic #4181 / openjdk23_java-matrix](https://gradle-enterprise.elastic.co/s/sv7jhxpueldic)
- [elasticsearch-periodic #4173 / openjdk23_java-matrix](https://gradle-enterprise.elastic.co/s/5dc6jsemskmp2)
- [elasticsearch-periodic #4145 / openjdk23_java-matrix](https://gradle-enterprise.elastic.co/s/7ailzwhmx2fks)
- [elasticsearch-periodic #4137 / openjdk23_java-matrix](https://gradle-enterprise.elastic.co/s/2m2rwpahdibyu)
- [elasticsearch-periodic #4133 / openjdk23_java-matrix](https://gradle-enterprise.elastic.co/s/ezkfazynb5lew)
- [elasticsearch-periodic #4129 / openjdk23_java-matrix](https://gradle-enterprise.elastic.co/s/qqgonijbzkvsq)

**Reproduction Line:**
```
./gradlew ':modules:repository-url:test' --tests "org.elasticsearch.common.blobstore.url.URLBlobContainerRetriesTests.testReadBlobWithPrematureConnectionClose" -Dtests.seed=D7D311EF0CD4361A -Dtests.locale=fr-CM -Dtests.timezone=Asia/Aqtobe -Druntime.java=23
```

**Applicable branches:**
7.17

**Reproduces locally?:**
N/A

**Failure History:**
[See dashboard](https://es-delivery-stats.elastic.dev/app/dashboards#/view/dcec9e60-72ac-11ee-8f39-55975ded9e63?_g=(refreshInterval:(pause:!t,value:60000),time:(from:now-7d%2Fd,to:now))&_a=(controlGroupInput:(chainingSystem:HIERARCHICAL,controlStyle:twoLine,ignoreParentSettings:(ignoreFilters:!f,ignoreQuery:!f,ignoreTimerange:!f,ignoreValidations:!t),panels:('0c0c9cb8-ccd2-45c6-9b13-96bac4abc542':(explicitInput:(dataViewId:fbbdc689-be23-4b3d-8057-aa402e9ed0c5,enhancements:(),fieldName:task.keyword,grow:!t,id:'0c0c9cb8-ccd2-45c6-9b13-96bac4abc542',searchTechnique:wildcard,selectedOptions:!(),singleSelect:!t,title:'Gradle%20Task',width:medium),grow:!t,order:0,type:optionsListControl,width:small),'144933da-5c1b-4257-a969-7f43455a7901':(explicitInput:(dataViewId:fbbdc689-be23-4b3d-8057-aa402e9ed0c5,enhancements:(),fieldName:name.keyword,grow:!t,id:'144933da-5c1b-4257-a969-7f43455a7901',searchTechnique:wildcard,selectedOptions:!('testReadBlobWithPrematureConnectionClose'),title:Test,width:medium),grow:!t,order:2,type:optionsListControl,width:medium),'4e6ad9d6-6fdc-4fcc-bf1a-aa6ca79e0850':(explicitInput:(dataViewId:fbbdc689-be23-4b3d-8057-aa402e9ed0c5,enhancements:(),fieldName:className.keyword,grow:!t,id:'4e6ad9d6-6fdc-4fcc-bf1a-aa6ca79e0850',searchTechnique:wildcard,selectedOptions:!('org.elasticsearch.common.blobstore.url.URLBlobContainerRetriesTests'),title:Suite,width:medium),grow:!t,order:1,type:optionsListControl,width:medium)))))

**Failure Message:**
```
java.lang.AssertionError: 
Expected: ((a string containing "premature end of chunk coded message body: closing chunk expected" or a string containing "premature end of content-length delimited message body") or a string containing "connection closed prematurely")
     but: was "the target server failed to respond"
```

**Issue Reasons:**
- [7.17] 9 failures in test testReadBlobWithPrematureConnectionClose (1.5% fail rate in 615 executions)
- [7.17] 9 failures in step openjdk23_java-matrix (52.9% fail rate in 17 executions)
- [7.17] 9 failures in pipeline elasticsearch-periodic (47.4% fail rate in 19 executions)

**Note:**
This issue was created using new test triage automation. Please report issues or feedback to es-delivery.

### PR DESCRIPTION
**Pull request by DaveCTurner**
It seems that JDK23 yields a different error message if the connection
is closed before even sending headers. This commit relaxes the assertion
to match, and also tightens it up to check that we report a premature
end of the body if we _do_ send headers.

Closes #113595

### CODE REVIEW

File: test/framework/src/main/java/org/elasticsearch/repositories/blobstore/AbstractBlobContainerRetriesTestCase.java [hunk 1]
```diff
@@ -341,9 +348,12 @@ public void testReadBlobWithPrematureConnectionClose() {
...
         assertThat(
             exception.getMessage().toLowerCase(Locale.ROOT),
-            either(containsString("premature end of chunk coded message body: closing chunk expected")).or(
-                containsString("premature end of content-length delimited message body")
-            ).or(containsString("connection closed prematurely"))
+            anyOf(
+                containsString("premature end of chunk coded message body: closing chunk expected"),
+                containsString("premature end of content-length delimited message body"),
+                containsString("connection closed prematurely"),
+                alwaysFlushBody ? never() : containsString("the target server failed to respond")
```
Comment (DiannaHohensee) [inline test/framework/src/main/java/org/elasticsearch/repositories/blobstore/AbstractBlobContainerRetriesTestCase.java:355]: Could you add a comment here?

It looks like if we flush the output, we expect the original errors. Without flushing the output stream, then we always get a standard error message that the node is gone?

**Review by DiannaHohensee — COMMENTED** [review]
Where is the error /stack trace in the test failure (https://github.com/elastic/elasticsearch/issues/113595)?

**Comment by DaveCTurner** [comment]
> Where is the error /stack trace in the test failure

[here](https://gradle-enterprise.elastic.co/s/ox2rdcz4v4kc6/tests/task/:modules:repository-url:test/details/org.elasticsearch.common.blobstore.url.URLBlobContainerRetriesTests/testReadBlobWithPrematureConnectionClose?expanded-stacktrace=WyIwIl0&top-execution=1):

```
java.lang.AssertionError:   
Expected: ((a string containing "premature end of chunk coded message body: closing chunk expected" or a string containing "premature end of content-length delimited message body") or a string containing "connection closed prematurely")    
     but: was "the target server failed to respond" 
    at __randomizedtesting.SeedInfo.seed([D7D311EF0CD4361A:233B4CEF6B8E0634]:0) 
    at org.hamcrest.MatcherAssert.assertThat(MatcherAssert.java:18) 
    at org.junit.Assert.assertThat(Assert.java:964) 
    at org.junit.Assert.assertThat(Assert.java:930) 
    at org.elasticsearch.repositories.blobstore.AbstractBlobContainerRetriesTestCase.testReadBlobWithPrematureConnectionClose(AbstractBlobContainerRetriesTestCase.java:327)    
    at jdk.internal.reflect.DirectMethodHandleAccessor.invoke(DirectMethodHandleAccessor.java:103)  
    at java.lang.reflect.Method.invoke(Method.java:580) 
    at com.carrotsearch.randomizedtesting.RandomizedRunner.invoke(RandomizedRunner.java:1758)   
    at com.carrotsearch.randomizedtesting.RandomizedRunner$8.evaluate(RandomizedRunner.java:946)    
    at com.carrotsearch.randomizedtesting.RandomizedRunner$9.evaluate(RandomizedRunner.java:982)    
    at com.carrotsearch.randomizedtesting.RandomizedRunner$10.evaluate(RandomizedRunner.java:996)   
    at com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36) 
    at org.junit.rules.RunRules.evaluate(RunRules.java:20)  
    at org.apache.lucene.util.TestRuleSetupTeardownChained$1.evaluate(TestRuleSetupTeardownChained.java:49) 
    at org.apache.lucene.util.AbstractBeforeAfterRule$1.evaluate(AbstractBeforeAfterRule.java:45)   
    at org.apache.lucene.util.TestRuleThreadAndTestName$1.evaluate(TestRuleThreadAndTestName.java:48)   
    at org.apache.lucene.util.TestRuleIgnoreAfterMaxFailures$1.evaluate(TestRuleIgnoreAfterMaxFailures.java:64) 
    at org.apache.lucene.util.TestRuleMarkFailure$1.evaluate(TestRuleMarkFailure.java:47)   
    at org.junit.rules.RunRules.evaluate(RunRules.java:20)  
    at com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36) 
    at com.carrotsearch.randomizedtesting.ThreadLeakControl$StatementRunner.run(ThreadLeakControl.java:390) 
    at com.carrotsearch.randomizedtesting.ThreadLeakControl.forkTimeoutingTask(ThreadLeakControl.java:843)  
    at com.carrotsearch.randomizedtesting.ThreadLeakControl$3.evaluate(ThreadLeakControl.java:490)  
    at com.carrotsearch.randomizedtesting.RandomizedRunner.runSingleTest(RandomizedRunner.java:955) 
    at com.carrotsearch.randomizedtesting.RandomizedRunner$5.evaluate(RandomizedRunner.java:840)    
    at com.carrotsearch.randomizedtesting.RandomizedRunner$6.evaluate(RandomizedRunner.java:891)    
    at com.carrotsearch.randomizedtesting.RandomizedRunner$7.evaluate(RandomizedRunner.java:902)    
    at org.apache.lucene.util.AbstractBeforeAfterRule$1.evaluate(AbstractBeforeAfterRule.java:45)   
    at com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36) 
    at org.apache.lucene.util.TestRuleStoreClassName$1.evaluate(TestRuleStoreClassName.java:41) 
    at com.carrotsearch.randomizedtesting.rules.NoShadowingOrOverridesOnMethodsRule$1.evaluate(NoShadowingOrOverridesOnMethodsRule.java:40) 
    at com.carrotsearch.randomizedtesting.rules.NoShadowingOrOverridesOnMethodsRule$1.evaluate(NoShadowingOrOverridesOnMethodsRule.java:40) 
    at com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36) 
    at com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36) 
    at org.apache.lucene.util.TestRuleAssertionsRequired$1.evaluate(TestRuleAssertionsRequired.java:53) 
    at org.apache.lucene.util.TestRuleMarkFailure$1.evaluate(TestRuleMarkFailure.java:47)   
    at org.apache.lucene.util.TestRuleIgnoreAfterMaxFailures$1.evaluate(TestRuleIgnoreAfterMaxFailures.java:64) 
    at org.apache.lucene.util.TestRuleIgnoreTestSuites$1.evaluate(TestRuleIgnoreTestSuites.java:54) 
    at org.junit.rules.RunRules.evaluate(RunRules.java:20)  
    at com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36) 
    at com.carrotsearch.randomizedtesting.ThreadLeakControl$StatementRunner.run(ThreadLeakControl.java:390) 
    at com.carrotsearch.randomizedtesting.ThreadLeakControl.lambda$forkTimeoutingTask$0(ThreadLeakControl.java:850) 
    at java.lang.Thread.run(Thread.java:1575)       
```

File: test/framework/src/main/java/org/elasticsearch/repositories/blobstore/AbstractBlobContainerRetriesTestCase.java (hunk 1 above)
Comment (DaveCTurner) [reply test/framework/src/main/java/org/elasticsearch/repositories/blobstore/AbstractBlobContainerRetriesTestCase.java:360]: Does e6d486bd1c2173006477413d799f92870c4c6e7b help?

Comment (DiannaHohensee) [reply test/framework/src/main/java/org/elasticsearch/repositories/blobstore/AbstractBlobContainerRetriesTestCase.java:360]: Yep, thanks

**Review by DiannaHohensee — APPROVED** [review]
LGTM. 🚢

**Pull request merged**